Azure Functions docs (#8237)
Particular/docs.particular.net: nservicebus/hosting/azure/functions/index.md

  - nservicebus/hosting/override-hostid
  - samples/azure-functions/service-bus
---

NServiceBus endpoints can be hosted in an Azure Functions app using the [`NServiceBus.AzureFunctions.AzureServiceBus`](https://www.nuget.org/packages/NServiceBus.AzureFunctions.AzureServiceBus) package:

snippet: azure-functions-basic-endpoint

An endpoint is declared as a partial method inside a partial class and composed of two parts:

- A partial method decorated with a `[NServiceBusFunction]` and a `[Function("MyFunction")]` attribute declaring a [Azure Service Bus trigger](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus-trigger).
  - The trigger must set `AutoCompleteMessages = false` since NServiceBus pipeline takes responsibility for completing or abandoning each message based on handler outcomes; the `NSBFUNC006` analyzer enforces this requirement.
  - A source generator will emit the method body that forwards each incoming message to the NServiceBus pipeline.
- A static `Configure{FunctionName}` method that configures the NServiceBus endpoint for the function.

> [!NOTE]
> A single Functions app can host one or more endpoints.

> ## The configure method

Reviewer: I think this heading (the configure method), Transport Configuration, and "Send-only endpoints" could be moved to a separate page "configuring function endpoints"

| [`ThrowOnMissingTopicWhenPublishing`](/transports/azure-service-bus/configuration.md#entity-creation) | Controls whether publishing to a non-existent topic throws after logging a warning. |

Other receive-side Azure Service Bus transport settings, such as [prefetch count](/transports/azure-service-bus/configuration.md#controlling-the-prefetch-count) and [lock renewal](/transports/azure-service-bus/configuration.md#lock-renewal), are controlled by Azure Functions rather than by `AzureServiceBusServerlessTransport` in this hosting model.

## Explicit handler and saga registration

Assembly scanning is not available in this hosting model, so handlers and sagas must be registered explicitly. See [Registering Handlers and Sagas](/nservicebus/handlers-and-sagas-registration.md) for the available registration approaches in NServiceBus 10.2, and [Disable assembly scanning](/nservicebus/hosting/assembly-scanning.md#disable-assembly-scanning) for the caveats and additional manual registration requirements. Handlers that are not registered will not be invoked.

## Wiring the host

> The Functions host is bootstrapped in `Program.cs`. Calling `AddNServiceBusFunctions` registers and configures each `[NServiceBusFunction]` with the Functions runtime:

Reviewer: I think this needs to be mentioned at the top of the page so it's not buried where all this magic gets wired up.

There also needs to be more said about how this method is source generated and isn't available until there's function endpoint attributes detected in the compilation.

Reviewer 2: I've made the intro more terse now, take another look

The queue that the endpoint receives from is taken from the `queueName` passed to the `[ServiceBusTrigger]` attribute. No separate receive queue name is configured in `Configure{FunctionName}`.

The queue name in `[ServiceBusTrigger]` can use Azure Functions [binding expressions](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-expressions-patterns).

## Transactions

Azure Functions endpoints only supports [`TransportTransactionMode.ReceiveOnly`](/transports/transactions.md#transaction-modes-transport-transaction-receive-only).

## Recoverability

In addition to the standard recoverability provided by NServiceBus the package also automatically enables [DLQ forwarding to the error queue](/transports/azure-service-bus/configuration.md#dead-lettering-forward-dead-lettered-messages-to-the-error-queue), so dead-lettered messages can be managed alongside other failed messages by tools such as ServicePulse.

For details on the dead-letter behavior and configuration options, see [dead-lettering](/transports/azure-service-bus/configuration.md#dead-lettering) in the Azure Service Bus transport documentation.

## Provisioning the namespace

> Provision queues and other entities in the Azure Service Bus namespace using the [`asb-transport` command-line tool](/transports/azure-service-bus/operational-scripting.md).

Reviewer: This section makes it sound like EnableInstallers() won't work (but it does)

> [!WARNING]
> The Azure Service Bus transport defaults `MaxDeliveryCount` to 100. When provisioning entities with the `asb-transport` tool or by hand, set `MaxDeliveryCount` to match (for example, pass `--maximum-delivery-count 100` to the CLI tool) so the namespace and the transport agree. Also make sure to include dead-letter queue forwarding using `--forward-dlq-to error`

## Startup diagnostics

Startup diagnostics are not written by default in this hosting model. Use `LogDiagnostics()` to write them to the Functions host log stream and any configured destination such as Application Insights. For the available `LogDiagnostics()` behavior and options, see [Writing diagnostics to the log](/nservicebus/hosting/startup-diagnostics.md#writing-diagnostics-to-the-log).

### Custom diagnostics writer

For full control over the diagnostic output (for example, to persist diagnostics beyond the function's execution lifetime or to aggregate diagnostic data from multiple function instances), configure a [custom diagnostics writer](/nservicebus/hosting/startup-diagnostics.md) on the endpoint configuration.

> ## Analyzers

Reviewer: Separate page?

---
title: Azure Functions hosting
component: AzureFunctions
summary: Hosting NServiceBus endpoints in Azure Functions with the AzureServiceBus package
reviewed: 2026-06-02
related:
  - transports/azure-service-bus
  - nservicebus/hosting/startup-diagnostics
  - nservicebus/hosting/override-hostid
  - samples/azure-functions/service-bus
---

NServiceBus endpoints can be hosted in an Azure Functions app using the [`NServiceBus.AzureFunctions.AzureServiceBus`](https://www.nuget.org/packages/NServiceBus.AzureFunctions.AzureServiceBus) package:

snippet: azure-functions-basic-endpoint

An endpoint is declared as a partial method inside a partial class and composed of two parts:

> - A partial method is decorated with a `[NServiceBusFunction]` and a `[Function("MyFunction")]` attribute declaring a [Azure Service Bus trigger](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus-trigger).

Reviewer 3: ```suggestion
- A partial method decorated with a `[NServiceBusFunction]` and a `[Function("MyFunction")]` attribute declaring a [Azure Service Bus trigger](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus-trigger).
```

>   - A source generator will emits the method body that forwards each incoming message to the NServiceBus pipeline.

Reviewer 3: ```suggestion
  - A source generator will emit the method body that forwards each incoming message to the NServiceBus pipeline.
```

>   - The trigger must set `AutoCompleteMessages = false` since NServiceBus pipeline takes responsibility for completing or abandoning each message based on handler outcomes; the `NSBFUNC006` analyzer enforces this requirement.

Reviewer 3: Should this come first since it belongs to the trigger definition?

> - A static `Configure{FunctionName}` method that configures the NServiceBus endpoint for the function.

Reviewer 3: Should we show a pseudo code sample?

> [!NOTE]
> A single Functions app can host one or more endpoints.

## The configure method

> The static `Configure{FunctionName}` method is discovered by the source generator and called once per endpoint at host startup. Parameters are matched by type, in any combination:

Reviewer 3: In any combination? I thought we assumed endpoint config was always first...

| Parameter | Use |
|---|---|
| `EndpointConfiguration` | Required. Configures handlers, recoverability, audit, and other endpoint settings. |

> | `IServiceCollection` | Optional. Registers endpoint-scoped services. |

Reviewer 3: Link to MS docs in all these?

| `IConfiguration` | Optional. Reads application configuration. |
| `IHostEnvironment` | Optional. Inspects the hosting environment, for example to differentiate development and production. |

Declare only the parameters needed:

snippet: azure-functions-configure-with-services

### Transport configuration

> Azure Functions endpoints must use `AzureServiceBusServerlessTransport`. Other transport definitions are not supported in this hosting model.

Reviewer 3: ```suggestion
Azure Functions endpoints must use `AzureServiceBusServerlessTransport`. Other transport definitions are currently not supported in this hosting model.
```

Should we add an issue they can click through and mention they'd like XYZ support?

| [`AutoForwardDeadLetteredMessagesToErrorQueue`](/transports/azure-service-bus/configuration.md#dead-lettering-forward-dead-lettered-messages-to-the-error-queue) | Enabled by default. Applies to transport-created receive queues. |
| [`MaxDeliveryCount`](/transports/azure-service-bus/configuration.md#entity-creation) | Applies when the transport creates queues. In this hosting model, `AzureServiceBusServerlessTransport` defaults `MaxDeliveryCount` to `100`, not `int.MaxValue` like `AzureServiceBusTransport`. |
| [`ThrowOnMissingTopicWhenPublishing`](/transports/azure-service-bus/configuration.md#entity-creation) | Controls whether publishing to a non-existent topic throws after logging a warning. |

Other receive-side Azure Service Bus transport settings, such as [prefetch count](/transports/azure-service-bus/configuration.md#controlling-the-prefetch-count) and [lock renewal](/transports/azure-service-bus/configuration.md#lock-renewal), are controlled by Azure Functions rather than by `AzureServiceBusServerlessTransport` in this hosting model.

## Explicit handler and saga registration

> Assembly scanning is not available in this hosting model, so handlers and sagas must be registered explicitly. See [Registering Handlers and Sagas](/nservicebus/handlers-and-sagas-registration.md) for the available registration approaches in NServiceBus 10.2, and [Disable assembly scanning](/nservicebus/hosting/assembly-scanning.md#disable-assembly-scanning) for the caveats and additional manual registration requirements. Handlers that are not registered will not be invoked.

Reviewer 3: Could we rephrase this to be something like: To improve cold start time [assembly scanning](LINK) is disabled, therefore handlers and sagas must be registered...

Reviewer: Isn't the true cause, though, because a multi-endpoint hosting model can't do assembly scanning? "In order to support multiple NServiceBus endpoints within an Azure Functions application, function endpoints are automatically configured to disable assembly scanning…"

Could add the "this also provides the benefit of faster cold starts in the Functions environment" at the end but I think that's secondary.

## Wiring the host

The Functions host is bootstrapped in `Program.cs`. Calling `AddNServiceBusFunctions` registers and configures each `[NServiceBusFunction]` with the Functions runtime:

snippet: azure-functions-program-builder

> [!NOTE]

> > `AddNServiceBusFunctions` is a source-generated extension on `FunctionsApplicationBuilder` declared in the project's default namespace. Ensure `Program.cs` imports that namespace. At least one function needs to be defined for the namespace to be available.

Reviewer 3: Given we have an analyzer for this can or should this be removed or still worth keeping?

Reviewer 2: Removed

## Hosting multiple endpoints

Multiple methods decorated with `[NServiceBusFunction]` can co-exist in one Functions app. Each is registered as an independent NServiceBus endpoint with its own queue, handlers, and configure method. Two functions hosted from the same partial class:

snippet: azure-functions-multiple-endpoints

Each endpoint has its own `Configure{FunctionName}` method; the source generator routes each function to its matching configure method. Endpoints share the host's service provider but maintain independent message-handling pipelines.

## Send-only endpoints

A [send-only endpoint](/nservicebus/endpoints/#send-only) can be registered for components that need to dispatch messages without listening for incoming traffic, for example a function fronting an HTTP API:

snippet: azure-functions-sendonly-registration

> `AddSendOnlyNServiceBusEndpoint(...)` has overloads for endpoint-only configuration and for configuration plus endpoint-scoped service registration.

Reviewer 3: This probably has to change

> To send or publish from another function (for example, an HTTP trigger), inject `IMessageSession` keyed by the same name used at registration. Any endpoint-scoped services registered in the send-only endpoint can be resolved the same way:

Reviewer 3: Should we link to keyed services of MS docs?

snippet: azure-functions-sendonly-usage

The key passed to `[FromKeyedServices(...)]` must match the name passed to `AddSendOnlyNServiceBusEndpoint(...)`.

## Connection configuration

> The Azure Service Bus connection is read from the Functions configuration. The named setting can be either a [connection string](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-dotnet-get-started-with-queues#connection-string) or a configuration section containing `fullyQualifiedNamespace` for token-credential authentication. The same setting name is referenced in two places:

Reviewer 3: Isnt this mostly repeating how functions work in general and therefore more like MS documentation?

- The `Connection` parameter on the `[ServiceBusTrigger]` attribute.
- The `ConnectionName` property on `AzureServiceBusServerlessTransport` when registering a send-only endpoint.

> If the default setting name `AzureWebJobsServiceBus` is used, the `Connection` parameter can be omitted on `[ServiceBusTrigger]`, and `ConnectionName` does not need to be set on `AzureServiceBusServerlessTransport`.

Reviewer 3: Could we simplify this to "connection is inferred"?

The connection name in `[ServiceBusTrigger]` can also use Azure Functions [binding expressions](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-expressions-patterns).

## Queue name resolution

The queue that the endpoint receives from is taken from the `queueName` passed to the `[ServiceBusTrigger]` attribute. No separate receive queue name is configured in `Configure{FunctionName}`.

> The function name remains the logical endpoint name, while the integration uses the resolved trigger queue name as the local transport address when those values differ.

Reviewer 3: This is very technical... Not sure the user understands what this means

The queue name in `[ServiceBusTrigger]` can use Azure Functions [binding expressions](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-expressions-patterns).

## Transactions

> Azure Functions endpoints use [`TransportTransactionMode.ReceiveOnly`](/transports/transactions.md#transaction-modes-transport-transaction-receive-only). [`TransportTransactionMode.SendsAtomicWithReceive`](/transports/transactions.md#transaction-modes-transport-transaction-sends-atomic-with-receive) is not supported in this hosting model.

Reviewer 3: maybe we can say "and no other transport transaction mode is supported" because otherwise someone might ask "what about None"

## Recoverability

In addition to the standard recoverability provided by NServiceBus the package also automatically enables [DLQ forwarding to the error queue](/transports/azure-service-bus/configuration.md#dead-lettering-forward-dead-lettered-messages-to-the-error-queue), so dead-lettered messages can be managed alongside other failed messages by tools such as ServicePulse.

> For details on the dead-letter behavior and configuration options, see [dead-lettering](/transports/azure-service-bus/configuration.md#dead-lettering) in the Azure Service Bus transport documentation.

Reviewer 3: In the config overview, we repeated the config from ASB that was supported. Here we defer only. Should we be consistent?

## Provisioning the namespace

Provision queues and other entities in the Azure Service Bus namespace using the [`asb-transport` command-line tool](/transports/azure-service-bus/operational-scripting.md).

> [!WARNING]

> > The transport defaults `MaxDeliveryCount` to 100. When provisioning entities with the `asb-transport` tool or by hand, set `MaxDeliveryCount` to match (for example, pass `--maximum-delivery-count 100` to the CLI tool) so the namespace and the transport agree. Also make sure to include dead-letter queue forwarding using `--forward-dlq-to error`

Reviewer 3: Do we risk that the term "transport" is ambigious here?

---
title: Azure Functions hosting
component: AzureFunctions
summary: Hosting NServiceBus endpoints in Azure Functions with the AzureServiceBus package
reviewed: 2026-06-04
related:
  - transports/azure-service-bus
  - nservicebus/hosting/startup-diagnostics
  - nservicebus/hosting/override-hostid

>   - samples/azure-functions/service-bus

Reviewer: ```suggestion
  - samples/azure-functions/service-bus
  - nservicebus/upgrades/azure-functions-service-bus-isolated-to-azure-functions  
```

---

NServiceBus endpoints can be hosted in an Azure Functions app using the [`NServiceBus.AzureFunctions.AzureServiceBus`](https://www.nuget.org/packages/NServiceBus.AzureFunctions.AzureServiceBus) package.

> Integration is enabled in `Program.cs` by calling `AddNServiceBusFunctions`:

Reviewer: Except we generate this, so if you try to do this in an empty project, you won't find the method, right?

So I think we have 2 options:
1. Move this snippet after the endpoint snippet and call out the detail
2. Change the generator to an interceptor where the "baked in" method throws an exception with the message that either no endpoints were found or you need to turn source generators on.

Reviewer 2: I'd say #1 move it

Reviewer 3: Honestly the interceptor idea is pretty great and I think quite simple to migrate to given we are quite experienced in that already

Reviewer 2: So you are saying:

1. Add a AddNServiceBusFunctions() method that logs an error/warn (or do we throw?)
2. Put an attribute on it so we can easily intercept
3. Add an interceptor that replaces it with one that registers all the found functions IF we have at least one send-only or asb function found

I think that makes sense

Reviewer 3: I don't think we necessarily need an attribute. We already concluded for the core APIs the way we use interceptors is acceptable 

The rest sounds good. I think we don't necessarily need to throw or log. When you have nothing to intercept the method remains no-op. Logging that early is painful for us too because it happens in a static context

---
title: Azure Functions hosting
component: AzureFunctions
summary: Hosting NServiceBus endpoints in Azure Functions with the AzureServiceBus package
reviewed: 2026-05-06
---

NServiceBus endpoints can be hosted in an Azure Functions app using the [`NServiceBus.AzureFunctions.AzureServiceBus`](https://www.nuget.org/packages/NServiceBus.AzureFunctions.AzureServiceBus) package:

snippet: azure-functions-basic-endpoint

An endpoint is a partial class composed of three parts:

> - The class is decorated with `[NServiceBusFunction]`.

Reviewer 2: It can be on the trigger method as well so we need to decide which approach to lead with

## The configure method

The static `Configure{FunctionName}` method is discovered by the source generator and called once per endpoint at host startup. Parameters are matched by type, in any combination:

| Parameter | Use |
|---|---|
| `EndpointConfiguration` | Required. Configures handlers, recoverability, audit, and other endpoint settings. |
| `IServiceCollection` | Optional. Registers endpoint-scoped services. |
| `IConfiguration` | Optional. Reads application configuration. |
| `IHostEnvironment` | Optional. Inspects the hosting environment, for example to differentiate development and production. |

Declare only the parameters needed:

snippet: azure-functions-configure-with-services

## Handler registration

Handlers must be registered explicitly with `EndpointConfiguration.AddHandler<T>`. Assembly scanning is not available in this hosting model; handlers that are not registered will not be invoked.

## Wiring the host

The Functions host is bootstrapped in `Program.cs`. Calling `AddNServiceBusFunctions` discovers every `[NServiceBusFunction]` in the application and registers each endpoint with the Functions runtime:

snippet: azure-functions-program-builder

> [!NOTE]

> > `AddNServiceBusFunctions` is a source-generated extension on `FunctionsApplicationBuilder` declared in the project's default namespace. Ensure `Program.cs` imports that namespace.

Reviewer 2: ```suggestion
> `AddNServiceBusFunctions` is a source-generated extension on `FunctionsApplicationBuilder` declared in the project's default namespace. Ensure `Program.cs` imports that namespace. At least one function needs to be defined for the method to be available.
```

## Hosting multiple endpoints

Multiple `[NServiceBusFunction]`-decorated members can co-exist in one Functions app. Each is registered as an independent NServiceBus endpoint with its own queue, handlers, and configure method. Two functions hosted from the same partial class:

snippet: azure-functions-multiple-endpoints

Each endpoint has its own `Configure{FunctionName}` method; the source generator routes each function to its matching configure method. Endpoints share the host's service provider but maintain independent message-handling pipelines.

## Send-only endpoints

> A send-only endpoint can be registered for components that need to dispatch messages without listening for incoming traffic, for example a function fronting an HTTP API:

Reviewer 2: ```suggestion
A [send-only endpoint](/nservicebus/hosting#self-hosting-send-only-hosting) can be registered for components that need to dispatch messages without listening for incoming traffic, for example a function fronting an HTTP API:
```

snippet: azure-functions-sendonly-registration

To send or publish from another function (for example, an HTTP trigger), inject `IMessageSession` keyed by the same name used at registration:

snippet: azure-functions-sendonly-usage

The key passed to `[FromKeyedServices(...)]` must match the name passed to `AddSendOnlyNServiceBusEndpoint(...)`.

## Connection configuration

> The connection string for Azure Service Bus is read from the Functions configuration. The same setting name is referenced in two places:

Reviewer 2: ```suggestion
The [connection for Azure Service Bus](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus-trigger#connections) is read from the Functions configuration. The same setting name is referenced in two places:
```

- The `Connection` parameter on the `[ServiceBusTrigger]` attribute.
- The `ConnectionName` property on `AzureServiceBusServerlessTransport` when registering a send-only endpoint.

Any setting name can be used; the examples on this page use `ServiceBusConnection`.

## Logging and diagnostics

NServiceBus log output is forwarded to the Azure Functions logging pipeline automatically. Configure log levels and providers through the standard Functions logging configuration; no additional NServiceBus log setup is required.

> Startup diagnostics are written when the host starts. In Azure they appear in the Functions host log stream and in any configured destination such as Application Insights.

Reviewer 2: The default behavior is a no-op; writing to the log can be opted into via the new API @danielmarbach added in v10.2. Not sure if we have documented that yet?
